### PR TITLE
[Feature] Open student submission in online editor

### PR DESCRIPTION
**Pull request by fde312**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [x] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.
- [x] Client: I added multiple integration tests (Jest) related to the features
- [x] Client: I added `authorities` to all new routes and check the course groups for displaying navigation elements (links, buttons)
- [x] Client: I documented the TypeScript code using JSDoc style.
- [x] Client: I added multiple screenshots/screencasts of my UI changes

### Motivation and Context

Make it possible for tutors and instructors to review the code directly into Artemis instead of downloading the repo or checking it out via bitbucket.

### Description

- Use assessment-layout component for labouring
- Add code-editor container with individual nabber
- Reuse layout and functionality of dialog/modal component (which is now removed)

### Steps for Testing

1. Log in to Artemis
2. Navigate to Course Administration
3. Create programming exercise and participate in programming exercise (also exam mode)
4. Assess programming exercise (via tutorDashboard, exercise-score, examTutorDashboard)
4.1 Try out all buttons and check if something strange happens
5. Complain about assessment
6. Assess complaint
6.1 Try out all buttons and check if something strange happens

### Screenshots

New layout for the assessment of programming exercises:
image

image

Navbar elements are now below except automatic result (see screenshot 1)
image

Complaints:
Navbar elements are now below except automatic result (see screenshot 1)
image

image

### CODE REVIEW

**Comment by balazs-czopf** [comment]
Tested it on the testserver, nothing strange happened.

**Review by kloessst — COMMENTED** [review]
Already tested and didn't had any problems. Will do a code review tomorrow.

**Comment by TobiasPr** [comment]
Also joined in testing, could only find two things:

As a suggestion for the UI: Add some margin there, so that the elements do not touch each other:
image

Also when I override my assessment, the ui is not updated and I need to reload, see screenshot:
image

other than that, I did not find any issue.  I will also do a code review tomorrow

**Review by HanyaElhashemy — COMMENTED** [review]
Very good work. I find this a very useful feature!
I tested it for single exercises and everything (including complaining) looks good to me 👍 
As a student I can only see the complaint form if the assessment is manual, I think it is important to allow instructors to enable complaints for automatic results as well - just an idea for future work concerning programming exercises ^^

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+                    this.loadingParticipation = false;
+                },
+                () => {
+                    this.loadingParticipation = false;
+                },
+            );
+        });
+    }
+
+    private getLatestResult(results: Result[]): Result | null {
```
Comment (sascha11110) [inline src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:128]: Just a suggestion: first public, then private methods :-)

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+                return true;
+            }
+            if (this.complaint && this.isAssessor) {
+                // If there is a complaint, the original assessor cannot override the result anymore.
+                return false;
+            }
+            let isBeforeAssessmentDueDate = true;
+            // Add check as the assessmentDueDate must not be set for exercises
+            if (this.exercise.assessmentDueDate) {
+                isBeforeAssessmentDueDate = moment().isBefore(this.exercise.assessmentDueDate!);
```
Comment (sascha11110) [inline src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:257]: The exclamation mark at the end should not be necessary

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 3]
```diff
@@ -71,6 +89,114 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+            newResult.setSubmission(submission);
+            if (newResult.getSubmission() == null) {
+                throw new BadRequestAlertException("The submission is not connected to the result.", ENTITY_NAME, "submissionMissing");
+            }
+        }
+
+        Result result = programmingAssessmentService.saveManualAssessment(newResult);
+
+        if (submit) {
+
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:184]:
```suggestion
        if (submit) {
```

File: src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.html [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                [readOnly]="readOnly()"
+                name="resultString"
+                id="resultString"
+                [(ngModel)]="result.resultString"
+                (ngModelChange)="onResultModified.emit(result)"
+                #resultString="ngModel"
+            />
+            <span class="text-danger" *ngIf="resultString?.invalid && resultString?.errors?.required" jhiTranslate="artemisApp.result.resultStringRequired"
+                >Result text is required.</span
+            >
```
Comment (sascha11110) [inline src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.html:28]:
```suggestion
            <span class="text-danger" *ngIf="resultString?.invalid && resultString?.errors?.required" jhiTranslate="artemisApp.result.resultStringRequired">
                Result text is required.
            </span>
```

File: src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.html [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                (ngModelChange)="updateResultSuccess()"
+                #score="ngModel"
+            />
+            <span class="text-danger" *ngIf="score.dirty && score.invalid" jhiTranslate="artemisApp.result.invalidScore">Score is required and should be a number.</span>
+        </div>
+        <div class="checkbox col-md-6">
+            <label class="artemis-label">
+                <input type="checkbox" name="successful" [(ngModel)]="result.successful" [disabled]="true" />&nbsp;<span jhiTranslate="artemisApp.result.successful"
+                    >Successful</span
+                >
```
Comment (sascha11110) [inline src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.html:52]:
```suggestion
                <input type="checkbox" name="successful" [(ngModel)]="result.successful" [disabled]="true" />&nbsp;<span jhiTranslate="artemisApp.result.successful">
                    Successful
                </span>
```

File: src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.html [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+                    name="feedbackText{{ i }}"
+                    [readOnly]="readOnly()"
+                    required
+                    [(ngModel)]="feedback.text"
+                    (ngModelChange)="updateResultFeedbacks()"
+                    #feedbackText="ngModel"
+                />
+                <span class="text-danger" *ngIf="feedbackText?.invalid && feedbackText?.errors?.required" jhiTranslate="artemisApp.result.feedbackTextRequired"
+                    >Feedback text is required.</span
+                >
```
Comment (sascha11110) [inline src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.html:75]:
```suggestion
                <span class="text-danger" *ngIf="feedbackText?.invalid && feedbackText?.errors?.required" jhiTranslate="artemisApp.result.feedbackTextRequired">
                    Feedback text is required.
                </span>
```

**Review by sascha11110 — CHANGES_REQUESTED** [review]
Some comments on the code but nothing really critical :-)

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java (hunk 3 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:184]: done 😄

File: src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseStudentParticipationRepository.java [hunk 7]
```diff
@@ -24,6 +24,9 @@
     @Query("select p from ProgrammingExerciseStudentParticipation p left join fetch p.results pr left join fetch pr.feedbacks left join fetch pr.submission where p.id = :participationId and (pr.id = (select max(id) from p.results) or pr.id = null)")
     Optional<ProgrammingExerciseStudentParticipation> findByIdWithLatestResultAndFeedbacksAndRelatedSubmissions(@Param("participationId") Long participationId);
 
+    @Query("select p from ProgrammingExerciseStudentParticipation p left join fetch p.results pr left join fetch pr.feedbacks left join fetch pr.submission left join fetch pr.assessor where p.id = :participationId")
+    Optional<ProgrammingExerciseStudentParticipation> findByIdWithResultsAndFeedbacksAndRelatedSubmissions(@Param("participationId") Long participationId);
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/repository/ProgrammingExerciseStudentParticipationRepository.java:28]: done. I also renamed the method name in the Service :)

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java [hunk 8]
```diff
@@ -33,4 +40,55 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
         originalResult.setScore(assessmentUpdate.getScore());
         return super.updateAssessmentAfterComplaint(originalResult, exercise, assessmentUpdate);
     }
+
+    /**
+     * This function is used for saving a manual assessment/result. It sets the assessment type to MANUAL and sets the assessor attribute. Furthermore, it saves the result in the
+     * database.
+     *
+     // * @param submission the modeling submission to which the feedback belongs to
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:48]: jupe :)

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java [hunk 9]
```diff
@@ -33,4 +40,55 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
...
+     * @return result that was saved in the database
+     */
+    @Transactional
+    public Result saveManualAssessment(Result result) {
+        Boolean isFeedbackEmpty = result.getFeedbacks().isEmpty();
+        result.setHasFeedback(!isFeedbackEmpty);
+
+        User user = userService.getUserWithGroupsAndAuthorities();
+        // TODO: Double check why we set here complaint false
+        result.setHasComplaint(false);
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:59]: You are right, I forgot to remove my comment! :)

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java [hunk 10]
```diff
@@ -33,4 +40,55 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
...
+    @Transactional
+    public Result submitManualAssessment(long resultId) {
+        Result result = resultRepository.findWithEagerSubmissionAndFeedbackAndAssessorById(resultId)
+                .orElseThrow(() -> new EntityNotFoundException("No result for the given resultId could be found"));
+        // Every manual assessed programming submission is rated
+        result.setRated(true);
+        result.setCompletionDate(ZonedDateTime.now());
+
+        // Double calculatedScore = calculateTotalScore(result.getFeedbacks());
+        // return submitResult(result, exercise, calculatedScore);
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:90]: We will keep that, as we want to use this later on when we established to give points for manual results (similar to text and modeling) :)

[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:90]: I added a todo for this, so this is more clear why its commented out for the moment

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 11]
```diff
@@ -71,6 +89,114 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+            @RequestBody Result newResult) {
+        log.debug("REST request to save a new result : {}", newResult);
+        final var participation = participationService.findOneWithEagerResultsAndCourse(participationId);
+
+        User user = userService.getUserWithGroupsAndAuthorities();
+
+        Optional<Result> latestExistingResult = participation.getResults().stream().filter(result -> result.getAssessmentType() == AssessmentType.MANUAL).findFirst();
+        if (latestExistingResult.isPresent()) {
+            // prevent that tutors create multiple manual results
+            // TODO: Double check if this is the right way to do, or if we should create a new Manual Result
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:126]: Removed the comment! :)

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts (hunk 1 above)
Comment (fde312) [reply src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:128]: changed the order of methods :)

File: src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts (hunk 2 above)
Comment (fde312) [reply src/main/webapp/app/exercises/programming/assess/code-editor-tutor-assessment-container.component.ts:257]: done

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java [hunk 12]
```diff
@@ -33,4 +40,53 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
...
+    }
+
+    /**
+     * This function is used for submitting a manual assessment/result. It gets the result that belongs to the given resultId, updates the completion date, sets the assessment type
+     * to MANUAL and sets the assessor attribute. Afterwards, it saves the update result in the database again.
+     *
+     * @param resultId the id of the result that should be submitted
+     * @return the ResponseEntity with result as body
+     */
+    @Transactional
```
Comment (TobiasPr) [inline src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:79]: I am not that familiar with hibernate, but why are you using @Transactional here?

**Review by MaisiKoleni — APPROVED** [review]
Looks good to me.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 13]
```diff
@@ -71,6 +89,112 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+            // TODO: Double check if we can create a new result when the existing result has already a completion date (submitted)
+            // Create manual submission with last commit hash und current time stamp.
+            submission = programmingSubmissionService.createSubmissionWithLastCommitHashForParticipation((ProgrammingExerciseStudentParticipation) participation,
+                    SubmissionType.MANUAL);
+            newResult.setSubmission(submission);
+        }
+        else {
+            submission = programmingSubmissionService.findByIdWithEagerResultAndFeedback(latestExistingResult.get().getSubmission().getId());
+            newResult.setSubmission(submission);
+            if (newResult.getSubmission() == null) {
```
Comment (TobiasPr) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:175]: Wouldn't it be easier to check the submission directly for null, before setting the submission of the newResult and getting it again? Just for readability

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java (hunk 12 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:79]: To be honst, I am neither. I orientated myself to how it is done for the other exercise types. There, we used `@Transactional` see https://github.com/ls1intum/Artemis/blob/develop/src/main/java/de/tum/in/www1/artemis/service/ModelingAssessmentService.java#L79
that's why I used it as well to keep it consistent to the other exercises.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 14]
```diff
@@ -71,6 +89,114 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+        else if (newResult.getScore() < 100 && newResult.isSuccessful()) {
+            throw new BadRequestAlertException("Only result with score 100% can be successful.", ENTITY_NAME, "scoreAndSuccessfulNotMatching");
+        }
+        else if (!newResult.getFeedbacks().isEmpty() && newResult.getFeedbacks().stream().anyMatch(feedback -> feedback.getText() == null)) {
+            throw new BadRequestAlertException("In case feedback is present, feedback text and detail text are mandatory.", ENTITY_NAME, "feedbackTextOrDetailTextNull");
+        }
+
+        ProgrammingSubmission submission;
+        if (latestExistingResult.isEmpty()) {
+            // TODO: Double check if we can create a new result when the existing result has already a completion date (submitted)
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:166]: Done :)

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java (hunk 13 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:175]: Good catch! Fixed in

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 15]
```diff
@@ -71,6 +89,114 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+        checkAuthorization(exercise, user);
+
+        final var isAtLeastInstructor = authCheckService.isAtLeastInstructorForExercise(exercise, user);
+        if (!assessmentService.isAllowedToCreateOrOverrideResult(latestExistingResult.orElse(null), exercise, participation, user, isAtLeastInstructor)) {
+            log.debug("The user " + user.getLogin() + " is not allowed to override the assessment for the participation " + participation.getId() + " for User " + user.getLogin());
+            return forbidden("assessment", "assessmentSaveNotAllowed", "The user is not allowed to override the assessment");
+        }
+
+        final var course = exercise.getCourseViaExerciseGroupOrCourseMember();
+        if (!authCheckService.isAtLeastTeachingAssistantInCourse(course, user) || !exercise.areManualResultsAllowed()) {
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:144]: Good catch! Fixed in

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 16]
```diff
@@ -71,6 +89,114 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+
+        Result result = programmingAssessmentService.saveManualAssessment(newResult);
+
+        if (submit) {
+
+            result = programmingAssessmentService.submitManualAssessment(result.getId());
+        }
+        // remove information about the student for tutors to ensure double-blind assessment
+        if (!isAtLeastInstructor) {
+            ((StudentParticipation) result.getParticipation()).setParticipant(null);
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:187]: Thank you! I did not know that! Fixed in

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 17]
```diff
@@ -71,6 +89,114 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+        }
+        if (submit && ((result.getParticipation()).getExercise().getAssessmentDueDate() == null
+                || result.getParticipation().getExercise().getAssessmentDueDate().isBefore(ZonedDateTime.now()))) {
+            ltiService.onNewResult((ProgrammingExerciseStudentParticipation) result.getParticipation());
+            // TODO: Is it intended that the result with information about the assessor is broadcasted (this is also for other exercises type the case)
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:192]: Done :)

**Review by TobiasPr — APPROVED** [review]
Worked well in the testing session :+1:

**Review by MaisiKoleni — APPROVED** [review]

**Review by kloessst — APPROVED** [review]
Confirmed that everything works in multiple test sessions. Took a look at .ts and .java files (too many .html file got changed because of prettier) and it looks good to me.
Found just some small things.

File: src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java [hunk 18]
```diff
@@ -33,4 +40,53 @@ public Result updateAssessmentAfterComplaint(Result originalResult, Exercise exe
...
+     * This function is used for saving a manual assessment/result. It sets the assessment type to MANUAL and sets the assessor attribute. Furthermore, it saves the result in the
+     * database.
+     *
+     * @param result the new result of a programming exercise
+     * @return result that was saved in the database
+     */
+    @Transactional
+    public Result saveManualAssessment(Result result) {
+        Boolean isFeedbackEmpty = result.getFeedbacks().isEmpty();
+        result.setHasFeedback(!isFeedbackEmpty);
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/service/ProgrammingAssessmentService.java:54]: Done in

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 19]
```diff
@@ -71,6 +89,111 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+        else if (newResult.getScore() < 100 && newResult.isSuccessful()) {
+            throw new BadRequestAlertException("Only result with score 100% can be successful.", ENTITY_NAME, "scoreAndSuccessfulNotMatching");
+        }
+        else if (!newResult.getFeedbacks().isEmpty() && newResult.getFeedbacks().stream().anyMatch(feedback -> feedback.getText() == null)) {
+            throw new BadRequestAlertException("In case feedback is present, feedback text and detail text are mandatory.", ENTITY_NAME, "feedbackTextOrDetailTextNull");
+        }
+
+        ProgrammingSubmission submission;
+        if (latestExistingResult.isEmpty()) {
+            // Create manual submission with last commit hash und current time stamp.
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:165]: Done in

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java [hunk 20]
```diff
@@ -71,6 +89,111 @@ public ProgrammingAssessmentResource(AuthorizationCheckService authCheckService,
...
+            submission = programmingSubmissionService.createSubmissionWithLastCommitHashForParticipation((ProgrammingExerciseStudentParticipation) participation,
+                    SubmissionType.MANUAL);
+            newResult.setSubmission(submission);
+        }
+        else {
+            submission = programmingSubmissionService.findByIdWithEagerResultAndFeedback(latestExistingResult.get().getSubmission().getId());
+            if (submission == null) {
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingAssessmentResource.java:172]: you are right, very good catch! Done in

File: src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.ts [hunk 21]
```diff
@@ -291,6 +169,26 @@ export class ProgrammingAssessmentManualResultDialogComponent implements OnInit
...
         // at the moment instructors can still edit already accepted / rejected complaints because the first condition is true, however we do not yet allow to override complaints
-        return this.canOverride || (this.complaint !== undefined && this.complaint.accepted === undefined);
+        return this.canOverride || (this.complaint !== undefined && this.complaint.accepted === undefined && this.result.assessor.id !== this.user.id);
+    }
+    /**
+     * Updates if the result is successful (score of 100%) or not
+     * and emits the updated result to the parent component
+     */
+    updateResultSuccess() {
+        this.result.successful = this.result.score === 100;
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/webapp/app/exercises/programming/assess/manual-result/programming-assessment-manual-result-in-code-editor.component.ts:179]: Done, changed to >= 100 in

File: src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts [hunk 22]
```diff
@@ -11,7 +11,11 @@ import { ProgrammingExercise } from 'app/entities/programming-exercise.model';
 const BAMBOO_RESULT_LEGACY_TIMESTAMP = 1557526348000;
 
 export const isLegacyResult = (result: Result) => {
-    return result.completionDate!.valueOf() < BAMBOO_RESULT_LEGACY_TIMESTAMP;
+    if (result.completionDate) {
+        return result.completionDate.valueOf() < BAMBOO_RESULT_LEGACY_TIMESTAMP;
+    } else {
+        return true;
```
[the remark this answers is not in the document]
Comment (fde312) [reply src/main/webapp/app/exercises/programming/shared/utils/programming-exercise.utils.ts:17]: Very good catch! With this PR we introduce that results can also have no completionDate (when saving and but not yet submitted). Therefore this should be `false` :)
Change in

**Review by MaisiKoleni — APPROVED** [review]

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java [hunk 23]
```diff
@@ -76,6 +76,30 @@ public ProgrammingExerciseParticipationResource(ProgrammingExerciseParticipation
...
     }
 
+    /**
+     * Get the given student participation with its results and feedbacks.
+     *
+     * @param participationId for which to retrieve the student participation with results and feedbacks.
+     * @return the ResponseEntity with status 200 (OK) and the participation with its results in the body.
+     */
+    @GetMapping("/programming-exercise-participations/{participationId}/student-participation-with-results-and-feedbacks")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
```
Comment (krusche) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:86]: why should a user be allowed to call this resource?

**Review by krusche — CHANGES_REQUESTED** [review]
I think one Rest call is "too open" and should not be allowed for users

File: src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java (hunk 23 above)
Comment (fde312) [reply src/main/java/de/tum/in/www1/artemis/web/rest/ProgrammingExerciseParticipationResource.java:86]:
```suggestion
    @PreAuthorize("hasAnyRole('TA', 'INSTRUCTOR', 'ADMIN')")
```

**Review by TobiasPr — APPROVED** [review]

**Review by kloessst — APPROVED** [review]

**Review by MaisiKoleni — APPROVED** [review]

**Review by balazs-czopf — APPROVED** [review]

**Pull request merged**